File: Factory/PizzaStore.CS/PizzaStore/Ingredients.cs
namespace PizzaStore;


// INGREDIENT BASE CLASSES

public abstract class Dough
{
    public abstract String Name { get; set; }
}

public abstract class Sauce
{
    public abstract String Name { get; set; }
}

public abstract class Cheese
{
    public abstract String Name { get; set; }
}

public abstract class Veggies
{
    public abstract String Name { get; set; }
}

public abstract class Pepperoni
{
    public abstract String Name { get; set; }
}

public abstract class Clams
{
    public abstract String Name { get; set; }
}

// CONCRETE INGREDIENTS

public class ThinCrustDough : Dough
{
    public override String Name { get; set; } = "Thin Crust Dough";
}

public class MarinaraSauce : Sauce
{
    public override String Name { get; set; } = "Marinara Sauce";
}

public class ReggianoCheese : Cheese
{
    public override String Name { get; set; } = "Reggiano Cheese";
}

public class Garlic : Veggies
{
    public override String Name { get; set; } = "Garlic";
}

public class Onion : Veggies
{
    public override String Name { get; set; } = "Onion";
}

public class Mushroom : Veggies
{
    public override String Name { get; set; } = "Mushroom";
}

public class RedPepper : Veggies
{
    public override String Name { get; set; } = "RedPepper";
}

public class SlicedPepperoni : Pepperoni
{
    public override String Name { get; set; } = "SlicedPepperoni";
}

public class FreshClams : Clams
{
    public override String Name { get; set; } = "FreshClams";
}


// INGREDIENT FACTORIES

public interface PizzaIngredientFactory
{
    Dough CreateDough();
    Sauce CreateSauce();
    Cheese CreateCheese();
    List<Veggies> CreateVeggies();
    Pepperoni CreatePepperoni();
    Clams CreateClam();
}

public class NYPizzaIngredientFactory : PizzaIngredientFactory
{
    public Dough CreateDough()
    {
        return new ThinCrustDough();
    }

    public Sauce CreateSauce()
    {
        return new MarinaraSauce();
    }

    public Cheese CreateCheese()
    {
        return new ReggianoCheese();
    }

    public List<Veggies> CreateVeggies()
    {
        List<Veggies> veggies = new();
        veggies.Add(new Garlic());
        veggies.Add(new Onion());
        veggies.Add(new Mushroom());
        veggies.Add(new RedPepper());
        return veggies;
    }

    public Pepperoni CreatePepperoni()
    {
        return new SlicedPepperoni();
    }

    public Clams CreateClam()
    {
        return new FreshClams();
    }
}

File: Factory/Python/src/Ingredients.py
from abc import ABC
from typing import Protocol


class Dough(ABC):
    name: str

class Sauce(ABC):
    name: str

class Cheese(ABC):
    name: str

class Veggies(ABC):
    name: str

class Pepperoni(ABC):
    name: str

class Clams(ABC):
    name: str

class ThinCrustDough(Dough):
    name: str = "Thin Crust Dough"

class MarinaraSauce(Sauce):
    name: str = "Marinara Sauce"

class ReggianoCheese(Cheese):
    name: str = "Reggiano Cheese"

class Garlic(Veggies):
    name: str = "Garlic"

class Onion(Veggies):
    name: str = "Onion"

class Mushroom(Veggies):
    name: str = "Mushroom"

class RedPepper(Veggies):
    name: str = "RedPepper"

class SlicedPepperoni(Pepperoni):
    name: str = "SlicedPepperoni"

class FreshClams(Clams):
    name: str = "FreshClams"


class PizzaIngredientFactory(Protocol):
    def create_dough(self) -> Dough:...
    def create_sauce(self) -> Sauce:...
    def create_cheese(self) -> Cheese:...
    def create_veggies(self) -> list[Veggies]:...
    def create_pepperoni(self) -> Pepperoni:...
    def create_clam(self) -> Clams:...


class NYPizzaIngredientFactory:
    def create_dough(self) -> Dough:
        return ThinCrustDough()

    def create_sauce(self) -> Sauce:
        return MarinaraSauce()

    def create_cheese(self) -> Cheese:
        return ReggianoCheese()

    def create_veggies(self) -> list[Veggies]:
        veggies: list[Veggies] = []
        veggies.append(Garlic())
        veggies.append(Onion())
        veggies.append(Mushroom())
        veggies.append(RedPepper())
        return veggies

    def create_pepperoni(self) -> Pepperoni:
        return SlicedPepperoni()

    def create_clam(self) -> Clams:
        return FreshClams()